libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1334,25 +1335,101 @@ string YulUtilFunctions::allocationFunction()
 	});
 }
 
-string YulUtilFunctions::allocateMemoryArrayFunction(ArrayType const& _type)
+string YulUtilFunctions::allocateMemoryArrayFunction(ArrayType const& _type, bool _shouldZero)
 {
 	solUnimplementedAssert(!_type.isByteArray(), "");
 
-	string functionName = "allocate_memory_array_" + _type.identifier();
+	string functionName = "allocate_" + string(_shouldZero ? "and_zero_" : "") + "memory_array_" + _type.identifier();
+
+	if (_type.baseType()->hasSimpleZeroValueInMemory())
+		return m_functionCollector.createFunction(functionName, [&]() {
+			return Whiskers(R"(
+				function <functionName>(length) -> memPtr {
+					let allocSize := <allocSize>(length)
+					memPtr := <alloc>(allocSize)
+					<?shouldZero>
+					// cheap way of zero-initializing a memory range
+					calldatacopy(memPtr, calldatasize(), allocSize)
+					</shouldZero>
+					<?dynamic>
+					mstore(memPtr, length)
+					</dynamic>
+				}
+			)")
+			("functionName", functionName)
+			("alloc", allocationFunction())
+			("allocSize", arrayAllocationSizeFunction(_type))
+			("shouldZero", _shouldZero)
+			("dynamic", _type.isDynamicallySized())
+			.render();
+		});
+	else
+	{
+		auto const* base = _type.baseType();
+		solUnimplementedAssert(base->sizeOnStack() == 1, "Stacksize not yet implemented!");
+		return m_functionCollector.createFunction(functionName, [&]() {
+			return Whiskers(R"(
+				function <functionName>(length) -> memPtr {
+					let allocSize := <allocSize>(length)
+					memPtr := <alloc>(allocSize)
+					<?shouldZero>
+					let firstSlot := memPtr
+					let dataSize := allocSize
+					<?dynamic>
+					firstSlot := add(firstSlot, 32)
+					dataSize := sub(dataSize, 32)
+					</dynamic>
+					let words := div(add(dataSize, 31), 32)
+					for {let i := 0} lt(i, words) { i := add(i, 1) } {
```

Comment:
Why not add 32 to i and remove the multiplication and the rounding of dataSize to words?

Reply:
Done